From linux-usb-owner@vger.kernel.org  Tue Oct  5 15:43:16 2010
From: Christophe Lebouc <christophex.lebouc@intel.com>
Subject: usb: langwell_udc: fix big file transfer issue.
To: linux-usb@vger.kernel.org
Date: Thu, 30 Sep 2010 15:34:36 +0100
Message-ID: <20100930143428.6994.79214.stgit@localhost.localdomain>

From: Christophe Lebouc <christophex.lebouc@intel.com>

This patch fixing the problem with large file transfers failing. Swap the read
order to avoid unexpected RX status.

Signed-off-by: Christophe Lebouc <christophex.lebouc@intel.com>
Signed-off-by: Hao Wu <hao.wu@intel.com>
Signed-off-by: Alan Cox <alan@linux.intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/langwell_udc.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/drivers/usb/gadget/langwell_udc.c
+++ b/drivers/usb/gadget/langwell_udc.c
@@ -2440,12 +2440,14 @@ static int process_ep_req(struct langwel
 	dev_vdbg(&dev->pdev->dev, "---> %s()\n", __func__);
 
 	for (i = 0; i < curr_req->dtd_count; i++) {
-		remaining_length = le16_to_cpu(curr_dtd->dtd_total);
-		actual -= remaining_length;
 
 		/* command execution states by dTD */
 		dtd_status = curr_dtd->dtd_status;
 
+		barrier();
+		remaining_length = le16_to_cpu(curr_dtd->dtd_total);
+		actual -= remaining_length;
+
 		if (!dtd_status) {
 			/* transfers completed successfully */
 			if (!remaining_length) {
